### PR TITLE
feat: whitelist implementation

### PR DESCRIPTION
**Pull request by antmat**
PTAL

### CODE REVIEW

File: hub.yaml [hunk 1]
```diff
@@ -69,3 +69,7 @@ locator:
 market:
   # marketplace gRPC endpoint, required
   endpoint: "8125721C2413d99a33E351e1F6Bb4e56b6b633FD@127.0.0.1:9095"
+
+whitelist:
+  url: "https://raw.githubusercontent.com/sonm-io/allowed-list/master/general_whitelist.json"
+  enabled: true
```
Comment (3Hren) [inline hub.yaml]: seems redundant. we can keep whitelist as a pointer instead, it's optional anyway.

File: insonmnia/hub/server.go [hunk 2]
```diff
@@ -476,6 +489,42 @@ func (h *Hub) startTask(ctx context.Context, request *structs.StartTaskRequest)
 	return reply, nil
 }
 
+func (h *Hub) checkImageWhitelist(registry string, image string, auth string) (bool, reference.Named, error) {
```
Comment (3Hren) [inline insonmnia/hub/server.go]: I strictly against making the entire method for a single field `whitelist`. Decompose into `Whitelist` interface.

File: hub.yaml (hunk 1 above)
Comment (antmat) [reply hub.yaml]: As we discussed enabled whitelist is default

File: insonmnia/hub/whitelist.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		if err != nil {
+			return false, nil, err
+		}
+		allowed, err := w.digestAllowed(ref.Name(), (string)(digestedRef.Digest()))
+		return allowed, ref, err
+	}
+	dockerClient, err := dc.NewEnvClient()
+	if err != nil {
+		return false, nil, err
+	}
```
Comment (noxiouz) [inline insonmnia/hub/whitelist.go]: defer dockerClient.Close(). Otherwise, it can lead to leaking of idle http connections

**Pull request merged**